In [21]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from IPython.display import display

In [22]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# Part 1: Load the dataset

In [23]:
# Load the given datasets
ckd_numerical = pd.read_csv('/content/drive/MyDrive/Colab/chronic_kidney_disease_numerical.csv')
ckd_categorical = pd.read_csv('/content/drive/MyDrive/Colab/chronic_kidney_disease_categorical.csv')

# Print the data
display(ckd_numerical)
display(ckd_categorical)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: From my understanding it seems that the dataset is a collection of medical information from patients suspected of having chronic kidney disease (CKD).

Features: These are patient attributes or measurements used to predict the presence of CKD. They are both numerical and categorical.

* Numerical: age, bp (blood pressure), bgr (blood glucose random),
* Categorical: rbc (red blood cells), pc (pus cell), htn (hypertension)



Labels: This is what we are trying to predict: whether or not a patient has CKD (Target column).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [24]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
ckd_numerical.duplicated(subset='unique_id').sum()


# Drop these duplicate rows
ckd_numerical.drop_duplicates(subset='unique_id', inplace=True)

# Print the Dataset
display(ckd_numerical)

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
ckd_categorical.duplicated(subset='unique_id').sum()
ckd_categorical.drop_duplicates(subset='unique_id', inplace=True)
display(ckd_categorical)



,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
396,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
397,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
398,492993,3.0,0.0,NaN,normal,notpresent,notpresent,yes,no,no,good,no,yes,ckd
399,273038,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

*Answer*: No there are different unique id present in both datasets as we can see that the numerical dataset has 390 rows while the categorical has 400. If we do a inner join then the columns that exists only in one database wont show in the merged table. If we do a outer join then the column of unique_id will be a NaN wherever it was not present in the original dataframe

In [25]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined = pd.merge(ckd_numerical, ckd_categorical, on='unique_id', how='outer')


#Print the combined dataset
display(combined)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,103570,24.0,80.0,125.0,NaN,NaN,136.0,3.5,15.4,43.0,5600.0,4.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,104675,65.0,80.0,215.0,133.0,2.5,NaN,NaN,13.2,41.0,NaN,NaN,2.0,1.0,normal,normal,present,notpresent,no,yes,no,good,no,no,ckd
3,105923,90.0,90.0,139.0,89.0,3.0,140.0,4.1,12.0,37.0,7900.0,3.9,1.0,0.0,NaN,normal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
4,107297,60.0,70.0,NaN,NaN,NaN,NaN,NaN,16.4,43.0,10800.0,5.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
395,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,11500.0,3.3,4.0,3.0,normal,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd
396,996076,57.0,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,10400.0,6.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
397,996412,70.0,60.0,120.0,40.0,0.5,140.0,4.6,16.0,43.0,4500.0,4.9,NaN,NaN,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
398,997629,35.0,NaN,99.0,30.0,0.5,135.0,4.9,15.4,48.0,5000.0,5.2,0.0,0.0,normal,normal,NaN,NaN,no,no,no,good,no,no,notckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [26]:
# Calculate the percentage of rows that contain atleast one missing value
rows_with_missing = combined.isnull().any(axis=1)

# Print %
display((rows_with_missing.sum() / len(combined)) * 100
)

# Drop these rows from the dataset
combined.dropna(inplace=True)

combined_copy = combined.copy() #For Part4

# Print the Dataset
display(combined)

61.75000000000001

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
6,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,8100.0,4.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,114717,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,9100.0,5.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
8,118191,42.0,70.0,93.0,32.0,0.9,143.0,4.7,16.6,43.0,7100.0,5.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
9,118596,74.0,60.0,88.0,50.0,0.6,147.0,3.7,17.2,53.0,6000.0,4.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
390,980291,62.0,70.0,122.0,42.0,1.7,136.0,4.7,12.6,39.0,7900.0,3.9,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
391,991031,30.0,80.0,89.0,42.0,0.5,139.0,5.0,16.7,52.0,10200.0,5.0,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
394,994377,71.0,70.0,81.0,18.0,0.8,145.0,5.0,14.7,44.0,9800.0,6.0,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
395,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,11500.0,3.3,4.0,3.0,normal,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd


## Part 3.4: Sort the dataset according to the Labels

In [27]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
combined.sort_values(by='Target', inplace=True)
combined.reset_index(drop=True, inplace=True)

# Print the dataset
display(combined)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,6700.0,3.9,4.0,0.0,normal,abnormal,present,notpresent,yes,no,no,poor,yes,yes,ckd
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
3,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1,1.0,3.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
4,353025,40.0,70.0,253.0,150.0,11.9,132.0,5.6,10.9,31.0,8800.0,3.4,3.0,4.0,normal,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,5500.0,4.8,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,7900.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,7700.0,6.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,6700.0,6.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [28]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded = pd.get_dummies(combined, columns=categorical_columns, dtype = 'int')


# Print the dataset
display(encoded)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0
1,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,6700.0,3.9,4.0,0.0,0,1,1,0,0,1,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
3,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1,1.0,3.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
4,353025,40.0,70.0,253.0,150.0,11.9,132.0,5.6,10.9,31.0,8800.0,3.4,3.0,4.0,0,1,0,1,1,0,1,0,0,1,0,1,1,0,0,1,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,5500.0,4.8,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,7900.0,6.4,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,7700.0,6.3,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,6700.0,6.5,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [29]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for column in numerical_columns:
    mean = encoded[column].mean()
    std = encoded[column].std()
    encoded = encoded[(encoded[column] <= mean + 3 * std) & (encoded[column] >= mean - 3 * std)]

# Print the dataset
display(encoded)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
5,397388,63.0,100.0,280.0,35.0,3.2,143.0,3.5,13.0,40.0,9800.0,4.2,2.0,2.0,0,1,0,1,1,0,0,1,0,1,1,0,0,1,1,0,1,0,1,0,1,0
7,438182,73.0,80.0,253.0,142.0,4.6,138.0,5.8,10.5,33.0,7200.0,4.3,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
8,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9,4.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
9,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4,4.0,2.0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,5500.0,4.8,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,7900.0,6.4,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,7700.0,6.3,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,6700.0,6.5,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [30]:
# Normalize the all Numerical Attributes in the dataset.
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
encoded[numerical_columns] = scaler.fit_transform(encoded[numerical_columns])

# Print the dataset
display(encoded)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,343710,0.810190,0.730553,1.146438,4.014305,1.690051,-0.959702,1.373024,-2.913221,-2.905503,0.519658,-2.301962,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
5,397388,0.936367,2.776102,3.520558,-0.225995,1.218114,0.448743,-1.303189,-0.607180,-0.590673,0.785483,-1.084138,2.0,2.0,0,1,0,1,1,0,0,1,0,1,1,0,0,1,1,0,1,0,1,0,1,0
7,438182,1.567253,0.730553,2.921481,3.789156,2.161987,-0.431535,2.317570,-1.694935,-1.603411,-0.366426,-0.962356,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
8,474407,-1.713353,1.753327,-0.317973,-0.038371,0.206821,-2.720258,-1.303189,-2.652160,-3.050180,1.937392,-1.449485,4.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
9,475200,0.999456,2.776102,0.924557,0.486976,3.914895,-0.079424,0.428478,-2.826200,-2.616150,-0.233513,-2.058397,4.0,2.0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,-1.776441,-0.292221,0.037036,0.111728,-0.265116,-0.959702,-0.830916,0.088983,-0.011966,-1.119597,-0.353444,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
149,203466,-1.650264,-1.314996,-0.539853,-0.863916,-0.130277,-0.431535,-0.043795,-0.389629,-0.301320,-0.056296,1.595075,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
150,482105,-0.262316,-1.314996,-0.584229,0.186777,-0.602214,-0.431535,-0.201219,0.263024,0.856096,-0.144905,1.473292,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
151,228849,0.620924,0.730553,-0.473289,0.336877,-0.130277,-0.079424,-1.303189,-0.172078,0.856096,-0.587946,1.716857,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: I feel that the unique_id column is not serving as a feature for the model as it is purely for a human sanity to differentiate each of the entry from others. It is safe to delete the column for training purposes.

In [31]:
#Remove that column
encoded.drop(columns=['unique_id'], inplace=True)

# Print the dataset
display(encoded)

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,0.810190,0.730553,1.146438,4.014305,1.690051,-0.959702,1.373024,-2.913221,-2.905503,0.519658,-2.301962,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
5,0.936367,2.776102,3.520558,-0.225995,1.218114,0.448743,-1.303189,-0.607180,-0.590673,0.785483,-1.084138,2.0,2.0,0,1,0,1,1,0,0,1,0,1,1,0,0,1,1,0,1,0,1,0,1,0
7,1.567253,0.730553,2.921481,3.789156,2.161987,-0.431535,2.317570,-1.694935,-1.603411,-0.366426,-0.962356,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
8,-1.713353,1.753327,-0.317973,-0.038371,0.206821,-2.720258,-1.303189,-2.652160,-3.050180,1.937392,-1.449485,4.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
9,0.999456,2.776102,0.924557,0.486976,3.914895,-0.079424,0.428478,-2.826200,-2.616150,-0.233513,-2.058397,4.0,2.0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,-1.776441,-0.292221,0.037036,0.111728,-0.265116,-0.959702,-0.830916,0.088983,-0.011966,-1.119597,-0.353444,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
149,-1.650264,-1.314996,-0.539853,-0.863916,-0.130277,-0.431535,-0.043795,-0.389629,-0.301320,-0.056296,1.595075,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
150,-0.262316,-1.314996,-0.584229,0.186777,-0.602214,-0.431535,-0.201219,0.263024,0.856096,-0.144905,1.473292,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
151,0.620924,0.730553,-0.473289,0.336877,-0.130277,-0.079424,-1.303189,-0.172078,0.856096,-0.587946,1.716857,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [32]:
# Export the dataframe to a new csv file
encoded.to_csv('cleaned_data.csv', index=False)

# Export the dataframe to a new json file
encoded.to_json('cleaned_data.json', orient='records')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own.

In this part of the assignment you will prompt 383GPT to explore these capabilities.

In [33]:
pd.set_option('display.max_columns', None)
display(combined_copy.head(15)) # We made a copy of the dataframe after cleaning and before encoding to use it here

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
6,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,8100.0,4.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,114717,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,9100.0,5.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
8,118191,42.0,70.0,93.0,32.0,0.9,143.0,4.7,16.6,43.0,7100.0,5.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
9,118596,74.0,60.0,88.0,50.0,0.6,147.0,3.7,17.2,53.0,6000.0,4.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
11,124304,69.0,70.0,264.0,87.0,2.7,130.0,4.0,12.5,37.0,9600.0,4.1,3.0,4.0,normal,abnormal,notpresent,notpresent,yes,yes,yes,good,yes,no,ckd
13,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,7000.0,5.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
14,129053,48.0,80.0,133.0,139.0,8.5,132.0,5.5,10.3,36.0,6200.0,4.0,4.0,0.0,abnormal,abnormal,notpresent,present,no,yes,no,good,yes,no,ckd
15,129654,52.0,80.0,125.0,22.0,1.2,139.0,4.6,16.5,43.0,4700.0,4.6,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
16,131563,29.0,80.0,83.0,49.0,0.9,139.0,3.3,17.5,40.0,9900.0,4.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

| unique_id | age | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc | al | su | rbc | pc | pcc | ba | htn | dm | cad | appet | pe | ane | Target |
|-----------|------|------|-------|-------|-----|-------|-----|------|-----|----------|------|----|----|---------|----------|---------|----------|------|----|------|-------|-----|------|--------|
| 102677 | 59.0 | 70.0 | 76.0 | 186.0 | 15.0| 135.0 | 7.6 | 7.1 | 22.0| 3800.0 | 2.1 | 3.0| 0.0| normal | abnormal | notpresent| notpresent| yes | no | no | poor | yes | yes | ckd |
| 109053 | 64.0 | 60.0 | 106.0 | 27.0 | 0.7 | 150.0 | 3.3 | 14.4 | 42.0| 8100.0 | 4.7 | 0.0| 0.0| normal | normal | notpresent| notpresent| no | no | no | good | no | no | notckd |
| 114717 | 41.0 | 80.0 | 122.0 | 25.0 | 0.8 | 138.0 | 5.0 | 17.1 | 41.0| 9100.0 | 5.2 | 0.0| 0.0| normal | normal | notpresent| notpresent| no | no | no | good | no | no | notckd |
| 118191 | 42.0 | 70.0 | 93.0 | 32.0 | 0.9 | 143.0 | 4.7 | 16.6 | 43.0| 7100.0 | 5.3 | 0.0| 0.0| normal | normal | notpresent| notpresent| no | no | no | good | no | no | notckd |
| 118596 | 74.0 | 60.0 | 88.0 | 50.0 | 0.6 | 147.0 | 3.7 | 17.2 | 53.0| 6000.0 | 4.5 | 0.0| 0.0| normal | normal | notpresent| notpresent| no | no | no | good | no | no | notckd |
| 124304 | 69.0 | 70.0 | 264.0 | 87.0 | 2.7 | 130.0 | 4.0 | 12.5 | 37.0| 9600.0 | 4.1 | 3.0| 4.0| normal | abnormal | notpresent| notpresent| yes | yes| yes | good | yes | no | ckd |
| 125587 | 37.0 | 60.0 | 109.0 | 47.0 | 1.1 | 141.0 | 4.9 | 15.0 | 48.0| 7000.0 | 5.2 | 0.0| 0.0| normal | normal | notpresent| notpresent| no | no | no | good | no | no | notckd |
| 129053 | 48.0 | 80.0 | 133.0 | 139.0 | 8.5 | 132.0 | 5.5 | 10.3 | 36.0| 6200.0 | 4.0 | 4.0| 0.0| abnormal| abnormal | notpresent| present | no | yes| no | good | yes | no | ckd |
| 129654 | 52.0 | 80.0 | 125.0 | 22.0 | 1.2 | 139.0 | 4.6 | 16.5 | 43.0| 4700.0 | 4.6 | 0.0| 0.0| normal | normal | notpresent| notpresent| no | no | no | good | no | no | notckd |
| 131563 | 29.0 | 80.0 | 83.0 | 49.0 | 0.9 | 139.0 | 3.3 | 17.5 | 40.0| 9900.0 | 4.7 | 0.0| 0.0| normal | normal | notpresent| notpresent| no | no | no | good | no | no | notckd |
| 137148 | 73.0 | 100.0| 295.0 | 90.0 | 5.6 | 140.0 | 2.9 | 9.2 | 30.0| 7000.0 | 3.2 | 3.0| 2.0| abnormal| abnormal | present | notpresent| yes | yes| yes | poor | no | no | ckd |
| 153227 | 60.0 | 90.0 | 490.0 | 95.0 | 2.7 | 131.0 | 3.8 | 11.5 | 35.0| 12000.0 | 4.5 | 3.0| 5.0| abnormal| normal | notpresent| present | yes | yes| no | good | no | no | ckd |
| 155117 | 68.0 | 60.0 | 125.0 | 41.0 | 1.1 | 139.0 | 3.8 | 17.4 | 50.0| 6700.0 | 6.1 | 0.0| 0.0| normal | normal | notpresent| notpresent| no | no | no | good | no | no | notckd |
| 155872 | 59.0 | 60.0 | 113.0 | 23.0 | 1.1 | 139.0 | 3.5 | 15.3 | 54.0| 6500.0 | 4.9 | 0.0| 0.0| normal | normal | notpresent| notpresent| no | no | no | good | no | no | notckd |
| 160040 | 48.0 | 110.0| 106.0 | 215.0 | 15.2| 120.0 | 5.7 | 8.6 | 26.0| 5000.0 | 2.5 | 3.0| 0.0| abnormal| normal | present | notpresent| yes | no | yes | good | no | yes | ckd |

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [35]:
# Code to rename all the columns in the dataset
column_mapping = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'bgr': 'Blood Glucose Random',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count',
    'rbcc': 'Red Blood Cell Count',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc': 'Red Blood Cells',
    'pc': 'Pus Cell',
    'pcc': 'Pus Cell Clumps',
    'ba': 'Bacteria',
    'htn': 'Hypertension',
    'dm': 'Diabetes Mellitus',
    'cad': 'Coronary Artery Disease',
    'appet': 'Appetite',
    'pe': ' pedal edema',
    'ane': 'anemia',
    'Target': 'Target'
}

# Replacing the column names
combined_copy.rename(columns=column_mapping, inplace=True)

# Print the dataset
display(combined_copy)

,unique_id,Age,Blood Pressure,Blood Glucose Random,Blood Urea,Serum Creatinine,Sodium,Potassium,Hemoglobin,Packed Cell Volume,White Blood Cell Count,Red Blood Cell Count,Albumin,Sugar,Red Blood Cells,Pus Cell,Pus Cell Clumps,Bacteria,Hypertension,Diabetes Mellitus,Coronary Artery Disease,Appetite,pedal edema,anemia,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
6,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,8100.0,4.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,114717,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,9100.0,5.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
8,118191,42.0,70.0,93.0,32.0,0.9,143.0,4.7,16.6,43.0,7100.0,5.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
9,118596,74.0,60.0,88.0,50.0,0.6,147.0,3.7,17.2,53.0,6000.0,4.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
390,980291,62.0,70.0,122.0,42.0,1.7,136.0,4.7,12.6,39.0,7900.0,3.9,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
391,991031,30.0,80.0,89.0,42.0,0.5,139.0,5.0,16.7,52.0,10200.0,5.0,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
394,994377,71.0,70.0,81.0,18.0,0.8,145.0,5.0,14.7,44.0,9800.0,6.0,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
395,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,11500.0,3.3,4.0,3.0,normal,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [36]:
# Group by 'Target' and count the occurrences
result = combined_copy.groupby('Target').size().reset_index(name='count')

# Display the result
print(result)

   Target  count
0     ckd     42
1  notckd    111


*Describe what the above code does here*

The code is basically displaying total number of entries for the 2 types of targets we have and the count of each of the group. Basically its grouping all values of the target column into the bucket with other values same as that one.
